### PR TITLE
Add n-best list and embedding output options

### PR DESCRIPTION
**Pull request by asaluja**
To: @guillaumekln 

What: I added the ability to print out the n-best list when using the `-n_best` flag, and also print out the embeddings of the input source text using the `-print_input_embeddings` flag.

### CODE REVIEW

File: translate.lua [hunk 1]
```diff
@@ -32,7 +32,6 @@ local function main()
   local opt = cmd:parse(arg)
 
   _G.logger = onmt.utils.Logger.new(opt.log_file, opt.disable_logs, opt.log_level)
-  onmt.utils.Cuda.init(opt)
```
Comment (guillaumekln) [inline translate.lua]: Why was this line removed?

File: translate.lua [hunk 2]
```diff
@@ -107,30 +110,35 @@ local function main()
...
+                  _G.logger:info('BEST HYP:')
+                end
+              end
 
               if #results[b].preds > 1 then
-                _G.logger:info('')
-                _G.logger:info('BEST HYP:')
+                _G.logger:info("[%.2f] %s", results[b].preds[n].score, sentence)
+                if n ~= 1 then -- for the n == 1 case, we already wrote out the hypothesis above
+                  outFile:write(sentence .. '\n')
```
Comment (guillaumekln) [inline translate.lua]: There is no need to have special cases for `n == 1` and `n ~= 1`. Just call `write` at the beginning of the loop for all `n`.

File: onmt/translate/Translator.lua [hunk 3]
```diff
@@ -18,7 +18,9 @@ local options = {
   {'-pre_filter_factor', 1, [[Optional, set this only if filter is being used. Before
                             applying filters, hypotheses with top `beamSize * preFilterFactor`
                             scores will be considered. If the returned hypotheses voilate filters,
-                            then set this to a larger value to consider more.]]}
+                            then set this to a larger value to consider more.]]},
+  {'-print_input_embeddings', false, [[Instead of generating target tokens conditional on
```
Comment (guillaumekln) [inline onmt/translate/Translator.lua]: I find this option name ambiguous mostly because embeddings usually refer to word embeddings. What about `dump_input_encoding`?

File: onmt/translate/Translator.lua [hunk 4]
```diff
@@ -167,84 +169,87 @@ function Translator:translateBatch(batch)
...
   local encStates, context = self.models.encoder:forward(batch)
-
-  -- Compute gold score.
-  local goldScore
-  if batch.targetInput ~= nil then
-    if batch.uneven then
-      self.models.decoder:maskPadding(batch.sourceSize, batch.sourceLength)
+  if self.opt.print_input_embeddings then
+    return encStates
+  else
```
Comment (guillaumekln) [inline onmt/translate/Translator.lua]: Could you end the `if` block right there and keep the indentation of the rest of the function? Just so your PR does not conflict with others.

File: onmt/translate/Translator.lua [hunk 5]
```diff
@@ -167,84 +169,87 @@ function Translator:translateBatch(batch)
...
 
   local encStates, context = self.models.encoder:forward(batch)
-
-  -- Compute gold score.
-  local goldScore
-  if batch.targetInput ~= nil then
-    if batch.uneven then
-      self.models.decoder:maskPadding(batch.sourceSize, batch.sourceLength)
+  if self.opt.print_input_embeddings then
+    return encStates
```
Comment (guillaumekln) [inline onmt/translate/Translator.lua]: `encStates` contains the final cell and hidden states of all layers. Don't you want to just return the last hidden state `encStates[#encStates]`?

File: onmt/translate/Translator.lua [hunk 6]
```diff
@@ -274,18 +279,32 @@ function Translator:translate(src, gold)
...
-
-      results[b].preds = {}
-      for n = 1, self.opt.n_best do
-        results[b].preds[n] = {}
-        results[b].preds[n].words = self:buildTargetWords(pred[b][n], src[indexMap[b]].words, attn[b][n])
-        results[b].preds[n].features = self:buildTargetFeatures(predFeats[b][n])
-        results[b].preds[n].attention = attn[b][n]
-        results[b].preds[n].score = predScore[b][n]
+      if self.opt.print_input_embeddings then
+        results[b] = encStates
```
Comment (guillaumekln) [inline onmt/translate/Translator.lua]: Each batch should have its own encoding. If `translateBatch` returns the last hidden state, it's just about doing:

```lua
results[b] = encStates[b]
```

**Comment by asaluja** [comment]
@guillaumekln thanks so much for your comments! I will address your comments. It also seems like changes to `Translator.lua` have caused some conflicts, so I will also rebase.  I'll try to get to it as soon as I can.

**Comment by asaluja** [comment]
@guillaumekln please take a look.  I've made the changes you requested (thanks for providing those suggestions, I'm not too familiar with the codebase and lua so that was helpful) and resolved merge conflicts. 

One quick question: `encStates[#encStates]` should return the final hidden state in the last layer, right?

**Comment by guillaumekln** [comment]
Looks good to me.

Yes, `encStates[#encStates]` is the final hidden state in the last layer.
